src/sentry/search/events/fields.py

```diff
@@ -150,76 +153,168 @@ def project_threshold_config_expression(organization_id, project_ids):
+                    "arrayElement",
+                    [
+                        [
+                            "array",
+                            [
+                                [
+                                    "tuple",
+                                    [
+                                        "'{}'".format(TRANSACTION_METRICS[config["metric"]]),
+                                        config["threshold"],
+                                    ],
+                                ]
+                                for config in project_threshold_configs
+                            ],
+                        ],
+                        [
+                            "indexOf",
+                            [
+                                [
+                                    "array",
+                                    [
+                                        ["toUInt64", [config["project_id"]]]
+                                        for config in project_threshold_configs
+                                    ],
+                                ],
+                                "project_id",
                             ],
-                            "project_id",
                         ],
                     ],
-                    0,
                 ],
             ],
-            ["tuple", [f"'{DEFAULT_PROJECT_THRESHOLD_METRIC}'", DEFAULT_PROJECT_THRESHOLD]],
+        ]
+        if project_threshold_configs
+        else ["tuple", [f"'{DEFAULT_PROJECT_THRESHOLD_METRIC}'", DEFAULT_PROJECT_THRESHOLD]]
+    )
+
+    if transaction_threshold_configs:
+        return [
+            "if",
             [
-                "arrayElement",
                 [
+                    "equals",
                     [
-                        "array",
                         [
+                            "indexOf",
                             [
-                                "tuple",
                                 [
-                                    "'{}'".format(TRANSACTION_METRICS[config["metric"]]),
-                                    config["threshold"],
+                                    "array",
+                                    [
+                                        [
+                                            "tuple",
+                                            [
+                                                ["toUInt64", [config["project_id"]]],
+                                                "'{}'".format(config["transaction"]),
+                                            ],
+                                        ]
+                                        for config in transaction_threshold_configs
+                                    ],
                                 ],
-                            ]
-                            for config in threshold_configs
+                                ["tuple", ["project_id", "transaction"]],
+                            ],
```

Comment:
You can actually add an alias for this `indexOf` call here, e.g. `"transaction_threshold_index"` and reference the alias in the second argument of `arrayElement` below to save quite a few bytes. The same applies to the project threshold query above.